### PR TITLE
DualView sync's on ExecutionSpaces

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
This is something EMPIRE is going to need eventually, probably Trilinos too. In the spirit of removing these globally fencing 2-arg deep copies, DualViews should allow for a user to call sync on an execution space, and use that execution space for any deep copies it invokes. There's some complications, that lead to the design being a bit screwy.

First, the sync code is huge, I want to avoid code duplication. In my head, the way to do this, conceptually looks like

```c++
template<class... Args>
void sync(Args... args){
  if(i_should){ 
    Kokkos::deep_copy(args..., d_view,h_view); 
  }
}
```

If somebody passes an ExecutionSpace, we use it, otherwise we don't, just like the default.

Problem 1: the existing sync method does tricks with a default argument to specialize based on whether the DualView is of const or nonconst type.

Solution 1: Okay, use a delegating method

```c++
template<class... Args>
void sync_impl(const std::true_type&, Args... args){
  if(i_should){ 
    Kokkos::deep_copy(args..., d_view,h_view); 
  }
}

template<typename ExecutionSpace>
void sync(const ExecutionSpace& exec, /** chthonic metaprogramming incantation, true branch */) {
  sync_impl(true_type{}, exec);
}

void sync(/** chthonic metaprogramming incantation, true branch */) {
  sync_impl(true_type{});
}

//false left off here, but handled the same way
```

Problem 2: @crtrott making work for me. We have this gnarly block to handle UVM goodness in DualView's sync

```c++
Kokkos::Impl::cuda_prefetch_pointer(
  Kokkos::Cuda(), d_view.data(),
 sizeof(typename t_dev::value_type) * d_view.span(), false);
```

This should use the execution space *if and only if* it's a CUDA instance. So I added some plumbing for that, Impl::get_cuda_space, which returns a Cuda Space if one is passed to it, if anything else is passed to it, it just construct a default one (exactly what we currently do).

I really don't know enough about C++ design patterns to say whether this thing is implemented well or terribly, to be honest.

edit: Also, made our default deep copy instance for these be the deep_copy stream. Possible small asynchrony benefits there (see: https://github.com/kokkos/kokkos/pull/3822#discussion_r593469006 )

### CODE REVIEW

**Comment by DavidPoliakoff** [comment]
Note: not doing this for resize, as that allocates, which is a sync, too. So no point in optimizing that (unless we get async allocations)

**Review by nliber — APPROVED** [review]

File: containers/src/Kokkos_DualView.hpp [hunk 1]
```diff
@@ -480,12 +498,8 @@ class DualView : public ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type> {
...
   ///   as modified, by calling the modify() method with the
   ///   appropriate template parameter.
-  template <class Device>
-  void sync(const typename std::enable_if<
-                (std::is_same<typename traits::data_type,
-                              typename traits::non_const_data_type>::value) ||
-                    (std::is_same<Device, int>::value),
-                int>::type& = 0) {
+  template <class Device, class... Args>
+  void sync_impl(const std::true_type&, Args... args) {
```
[the remark this answers is not in the document]
Comment (DavidPoliakoff) [reply containers/src/Kokkos_DualView.hpp]: Thanks, implemented. Feel pretty safe passing the pack by value, as it's an ExecutionSpace instance, tops

File: containers/src/Kokkos_DualView.hpp [hunk 2]
```diff
@@ -480,12 +498,9 @@ class DualView : public ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type> {
...
   ///   as modified, by calling the modify() method with the
   ///   appropriate template parameter.
-  template <class Device>
-  void sync(const typename std::enable_if<
-                (std::is_same<typename traits::data_type,
-                              typename traits::non_const_data_type>::value) ||
-                    (std::is_same<Device, int>::value),
-                int>::type& = 0) {
+  // deliberately passing args by value as they're used multiple times
+  template <class Device, class... Args>
```
Comment (nliber) [inline containers/src/Kokkos_DualView.hpp:502]: After discussing this with @jrmadsen, the safest (and likely better performance) way to receive a parameter pack being used multiple times is by const l-value reference.

The general problem is that the callee can take parameters by non-const reference and modify them (this is obvious in the case of passing unnamed temporaries by non-const r-value reference, but the problem is more general than that), which we really don't want if we are passing them multiple times.  Better to receive by const reference (and the distinction between const l-value reference and const r-value reference isn't worth doing) so that the callees can't modify their values (`const_cast` and `mutable` notwithstanding, of course).  Plus, copies are only made if the callee does it (either implicitly with pass by value or explicitly in their code).

In other words, it should be written as:
```
template <class Device, class... Args>
void sync_impl(std::true_type, Args const&... args) {
```

**Review by nliber — DISMISSED** [review]

File: containers/src/Kokkos_DualView.hpp (hunk 2 above)
Comment (DavidPoliakoff) [reply containers/src/Kokkos_DualView.hpp:502]: Thanks for the explanation, done via e198274

**Comment by DavidPoliakoff** [comment]
Don't merge this one, I did something very dumb

**Comment by DavidPoliakoff** [comment]
Christian and I discussed over Slack. There's subtlety here, and I was worried about other problems we might encounter. Specifically, assuming a UVM View, and somebody does

```
my_uvm_dual_view.sync_host(my_host_space_instance);
```

Conceptually, this could be broken. On the other hand, it's the moral equivalent of

```c++
Kokkos::deep_copy(some_host_space, uvm_view_1, uvm_view_2);
```

So I'm going to make an incremental change here, but otherwise this will be good to go

**Comment by DavidPoliakoff** [comment]
Assuming this passes CI, it's ready for a review

File: containers/src/Kokkos_DualView.hpp [hunk 3]
```diff
@@ -91,6 +91,25 @@ namespace Kokkos {
...
  */
+
+namespace Impl {
+
+#ifdef KOKKOS_ENABLE_CUDA
+
+inline const Kokkos::Cuda& get_cuda_space(const Kokkos::Cuda& in) { return in; }
+
+inline Kokkos::Cuda get_cuda_space() {
+  return Kokkos::Cuda(Kokkos::Impl::cuda_get_deep_copy_stream());
```
Comment (masterleinad) [inline containers/src/Kokkos_DualView.hpp:102]: Any reason not to return a default constructed execution space here (which might be slightly cheaper)?

File: containers/src/Kokkos_DualView.hpp [hunk 4]
```diff
@@ -91,6 +91,25 @@ namespace Kokkos {
...
+}
+
+template <typename Arg>
+inline Kokkos::Cuda get_cuda_space(Arg&) {
+  return Kokkos::Cuda(Kokkos::Impl::cuda_get_deep_copy_stream());
```
Comment (masterleinad) [inline containers/src/Kokkos_DualView.hpp:107]: ... and here?

File: containers/src/Kokkos_DualView.hpp [hunk 5]
```diff
@@ -480,12 +500,9 @@ class DualView : public ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type> {
...
   ///   appropriate template parameter.
-  template <class Device>
-  void sync(const typename std::enable_if<
-                (std::is_same<typename traits::data_type,
-                              typename traits::non_const_data_type>::value) ||
-                    (std::is_same<Device, int>::value),
-                int>::type& = 0) {
+  // deliberately passing args by value as they're used multiple times
+  template <class Device, class... Args>
+  void sync_impl(const std::true_type&, Args const&... args) {
```
Comment (masterleinad) [inline containers/src/Kokkos_DualView.hpp:505]: What was the reason to not pass `std::true_type` by value? Also, you are passing `args` by reference which doesn't match the comment above anymore.

File: containers/src/Kokkos_DualView.hpp (hunk 3 above)
Comment (DavidPoliakoff) [reply containers/src/Kokkos_DualView.hpp:102]: So, if we did a default constructed space, we're enqueueing in the default stream, meaning that things in the default stream have to wait on these operations. This way, these copies don't interfere.

File: containers/src/Kokkos_DualView.hpp (hunk 4 above)
Comment (DavidPoliakoff) [reply containers/src/Kokkos_DualView.hpp:107]: See https://github.com/kokkos/kokkos/pull/3822#discussion_r590649342

File: containers/src/Kokkos_DualView.hpp (hunk 5 above)
Comment (DavidPoliakoff) [reply containers/src/Kokkos_DualView.hpp:505]: Good call, let me update those

File: containers/src/Kokkos_DualView.hpp (hunk 3 above)
Comment (masterleinad) [reply containers/src/Kokkos_DualView.hpp:102]: OK.

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

File: containers/src/Kokkos_DualView.hpp (hunk 5 above)
Comment (DavidPoliakoff) [reply containers/src/Kokkos_DualView.hpp]: fe6ef47

**Comment by DavidPoliakoff** [comment]
Retest this please

**Comment by DavidPoliakoff** [comment]
Retest this please.

**Comment by masterleinad** [comment]
You probably have to rebase to make this pass...

File: containers/src/Kokkos_DualView.hpp [hunk 6]
```diff
@@ -557,7 +590,27 @@ class DualView : public ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type> {
...
+  }
+  template <class Device, class ExecutionSpace>
+  void sync(const ExecutionSpace& exec,
+            const typename std::enable_if<
+                (!std::is_same<typename traits::data_type,
+                               typename traits::non_const_data_type>::value) ||
+                    (std::is_same<Device, int>::value),
+                int>::type& = 0) {
+    sync_impl<Device>(std::false_type{}, exec);
+  }
```
Comment (dalg24) [inline containers/src/Kokkos_DualView.hpp:609]: Surely you meant
```suggestion
  template <class Device>
  void sync() {
    sync_impl<Device>(std::intergal_constant<bool,
    std::is_same<typename traits::data_type, typename traits::non_const_data_type>::value ||
    std::is_same<Device, int>::value>);
  }
```

Comment (dalg24) [reply containers/src/Kokkos_DualView.hpp:609]: I don't understand why you mixSFINAE and tag dispatching

Comment (DavidPoliakoff) [reply containers/src/Kokkos_DualView.hpp:609]: Honestly, I mean that this should be completely rewritten, with this type trait made a `constexpr static bool` of the class. If you'd like to make this change I can look at it, though I know that as is it wouldn't compile. Also, I don't know what it means for the Device to be an int.

Comment (DavidPoliakoff) [reply containers/src/Kokkos_DualView.hpp:609]: Also, wait, your example doesn't pass the execution space, if we're given one?

Comment (dalg24) [reply containers/src/Kokkos_DualView.hpp:609]: I missed that

File: core/src/Kokkos_CudaSpace.hpp [hunk 7]
```diff
@@ -277,6 +277,8 @@ class CudaHostPinnedSpace {
 namespace Kokkos {
 namespace Impl {
 
+cudaStream_t cuda_get_deep_copy_stream();
```
[the remark this answers is not in the document]
Comment (DavidPoliakoff) [reply core/src/Kokkos_CudaSpace.hpp:280]: Done

File: containers/src/Kokkos_DualView.hpp (hunk 6 above)
Comment (dalg24) [reply containers/src/Kokkos_DualView.hpp:609]:
```C++
  template <class Device, class.. ExecSpaceOrNothing>
  void sync(ExecSpaceOrNothing const& exec_space_or_nothing...) {
    using Tag = std::intergal_constant<bool,
        std::is_same<typename traits::data_type, typename traits::non_const_data_type>::value ||
        std::is_same<Device, int>::value>;
    sync_impl<Device>(Tag{}, exec_space_or_nothing...);
  }

Comment (dalg24) [reply containers/src/Kokkos_DualView.hpp:609]: and you can try a
```C++
using Tuple = std::conditional_t<sizeof...(ExecSpaceOrNothing)==1, std::tuple< ExecSpaceOrNothing...>, std::tuple<void>>;
static_assert(sizeof...(ExecSpaceOrNothing)==0 || sizeof...(ExecSpaceOrNothing)==1 && Kokkos::is_execution_space<std::tuple_element_t<0,Tuple>>::value, "")

Comment (dalg24) [reply containers/src/Kokkos_DualView.hpp:609]: or better if you want make Daisy proud
```C++
#ifdef KOKKOS_ENABLE_CXX17
static_assert((... && Kokkos::is_execution_space<ExecSpaceOrNothing>::value));
#endif

Comment (DavidPoliakoff) [reply containers/src/Kokkos_DualView.hpp:609]: So, what I worry about here is that in the current version, we have functions with the same signatures I have provided, that is

```c++
template<typename Device> void sync(int)
```

So I am _positive_ that any program using DualView won't see changes. I can argue to myself that your change works: the only thing I can see breaking is if somebody incorrectly actually passed an `int` argument:

```c++
Kokkos::DualView<float*, Space> my_dv;
my_dv.sync<Space>(0);
```

This would be incorrect, but parses and works in the current version. That's just off the top of my head, and this is C++, there's always another corner case. If you think your edit needs to happen, I'll do it, but I worry about doing the subtle and clever thing to save ~30 lines, rather than guaranteeing ourselves the same interface. My recommendation is to have an action item for 4.0 to make this change, but emphasize stability here

**Review by crtrott — APPROVED** [review]

File: containers/src/Kokkos_DualView.hpp (hunk 6 above)
Comment (DavidPoliakoff) [reply containers/src/Kokkos_DualView.hpp:609]: Ah, thought of another: if you've taken a (member) function pointer to this function, this would be a breaking change. Of course I hope nobody is taking function pointers to these things, but I've hoped for a lot of things...

File: containers/src/Kokkos_DualView.hpp [hunk 8]
```diff
@@ -91,6 +91,25 @@ namespace Kokkos {
...
+
+#ifdef KOKKOS_ENABLE_CUDA
+
+inline const Kokkos::Cuda& get_cuda_space(const Kokkos::Cuda& in) { return in; }
+
+inline Kokkos::Cuda get_cuda_space() {
+  return Kokkos::Cuda(Kokkos::Impl::cuda_get_deep_copy_stream());
+}
+
+template <typename Arg>
```
Comment (DavidPoliakoff) [inline containers/src/Kokkos_DualView.hpp:105]: -> NonCudaExecSpace
and
`const`
